runtime/ms-rest-azure/lib/login.js

```diff
@@ -468,4 +470,232 @@ exports.withServicePrincipalSecretWithAuthResponse = function withServicePrincip
+  }
+  if (!credsObj.sqlManagementEndpointUrl) {
+    throw new Error(`"sqlManagementEndpointUrl" is missing from the auth file: ${filePath}.`);
+  }
+  if (!credsObj.galleryEndpointUrl) {
+    throw new Error(`"galleryEndpointUrl" is missing from the auth file: ${filePath}.`);
+  }
+  if (!credsObj.managementEndpointUrl) {
+    throw new Error(`"managementEndpointUrl" is missing from the auth file: ${filePath}.`);
+  }
+}
+
+function _foundManagementEndpointUrl(authFileUrl, envUrl) {
+  if (!authFileUrl || (authFileUrl && typeof authFileUrl.valueOf() !== 'string')) {
+    throw new Error('authFileUrl cannot be null or undefined and must be of type string.');
+  }
+
+  if (!envUrl || (envUrl && typeof envUrl.valueOf() !== 'string')) {
+    throw new Error('envUrl cannot be null or undefined and must be of type string.');
+  }
+
+  authFileUrl = authFileUrl.endsWith('/') ? authFileUrl.slice(0, -1) : authFileUrl;  
+  envUrl = envUrl.endsWith('/') ? envUrl.slice(0, -1) : envUrl;
+  return (authFileUrl.toLowerCase() === envUrl.toLowerCase());
+}
+
+/**
+ * Authenticates using the service principal information provided in the auth file. This method will set 
+ * the subscriptionId from the auth file to the user provided environment variable in the options 
+ * parameter or the default AZURE_SUBSCRIPTION_ID.
+ * @param {object} [options] - Optional parameters
+ * @param {string} [options.filePath] - Absolute file path to the auth file. If not provided 
+ * then please set the environment variable AZURE_AUTH_LOCATION.
+ * @param {string} [options.subscriptionEnvVariableName] - The subscriptionId environment variable 
+ * name. Default is 'AZURE_SUBSCRIPTION_ID'.
+ * @param {function} callback - The callback
+ */
+function _withAuthFile(options, callback) {
+  if (!callback && typeof options === 'function') {
+    callback = options;
+    options = {};
+  }
+  if (!options) options = { filePath: '' };
+  if (!callback) {
+    throw new Error('callback cannot be null or undefined.');
+  }
+
+  let filePath = options.filePath;
+  let authLocation = process.env[azureConstants.AZURE_AUTH_LOCATION];
+  let subscriptionEnvVariableName = options.subscriptionEnvVariableName || 'AZURE_SUBSCRIPTION_ID';
+  if (!filePath) {
+    if (!authLocation) {
+      let msg = `Either provide an absolute file path to the auth file or set/export the environment variable - ${azureConstants.AZURE_AUTH_LOCATION}.`;
+      return callback(new Error(msg));
+    }
+    filePath = authLocation;
+  }
+  //expand ~ to user's home directory.
+  if(filePath.startsWith('~')) {
+    filePath = msRest.homeDir(filePath.slice(1));
+  }
+
+  let content = null, credsObj = {}, optionsForSpSecret = {};
+  try {
+    content = fs.readFileSync(filePath, {encoding: 'utf8'});
+    credsObj = JSON.parse(content);
+    _validateAuthFileContent(credsObj, filePath);
+    if (!process.env[subscriptionEnvVariableName]) {
```

Comment:
Hope I understand the context correctly, the subscription from the auth file should be precedent the environment variable. You can overwrite the environment variable here, but that should be just an implementation detail being transparent to users.  Same with other fields. 
If I got it right, then we don't need an `if` condition, just overwrite it, optionally emit out a warning